### ISSUE TITLE
Question: listen to requests

### ISSUE THREAD

**Issue by brownad** [nfarina/homebridge#1085]
Hi,

Any examples on how homebridge / plugin can listen to an incoming request coming in, to then do something with it.

For example an endpoint on the network will make push a GET to homebridge with a specific query string based on a device doing something, how can homebridge listen for it on an Accessory, to then action off the back of it?  Anyone share how this would look?

**Comment by NorthernMan54** [comment]
I believe what you are asking is, if an event occurs at an accessory/device how can the device tell homebridge about the event, and can HomeKit update the device status with the updated status.  And this is feasible, and a couple of patterns exist.  The plugin could poll the device on regular intervals, and update the status, or the plugin could listen for events from the device, and update the status.

Both patterns are feasible and workable, but would require significant coding by the plugin developer.   If you look at my plugins, I have both patterns.  My homebridge-heyu plugin watches for events and updates HomeKit immediately, and homebridge-tcc plugin polls the thermostat for status every minute, and updates the status.

**Comment by brownad** [comment]
Nice one thanks

**Comment by brownad** [comment]
I got some of this working but went with the easy approach of polling continually. It fires off the first event but not subsequent ones. My approach is slightly wrong

Could you suggest how to set the state correctly and reset it please
https://github.com/brownad/homebridge-doorbird/blob/master/index.js

@northernman54

**Comment by NorthernMan54** [comment]
I was looking at your code, and is the issue around polling to event or is it with changing the status in homebridge?   At line 58, if you added a self.log line showing binaryState, does it fire correctly?  ( Before the IF )

Also, why are you using setTimeout?

Also, in my TCC plugin, I’m triggering the status update with this

 service.getCharacteristic(Characteristic.TargetTemperature).getValue();

And it causes getTemperature to execute, returning the new temperature

In your plugin, the same pattern would trigger

DoorBirdAccessory.prototype.getState

For this to work, binaryState would need to be stored in ‘this’ before getValue is called.

To summarize

polling to event watches for a change, and on change

Take the new binaryState and store it in ‘this' so that DoorBirdAccessory.prototype.getState
can use it to update status.

Then trigger DoorBirdAccessory.prototype.getState
with self.service.getCharacteristic(Characteristic.ProgrammableSwitchEvent).getValue()

The passing around of values with ‘this’ is a bit tricky, in TCC I’m using a global variable called accessory.

**Comment by brownad** [comment]
Thanks

I changed it albeit a rough approach to now update the state after it fires. It does work but looks bad in code. So I have it working.

I am suspicious about the notifications stability as it doesn't fire every time as expected I can't work that out. And it's also received on one iOS device and not the other two - any ideas how to push to other devices?

**Comment by brownad** [comment]
My setTimeout was incorrect, basically I wanted to stop receiving multiple push events to the user when state was '1' or you'll receive many IOS notifications.  I switched long polling on now about to test it.

Thanks for the feedback I need to go through it properly and clean up any suggestions would help

**Comment by snowdd1** [comment]
I made a "proof of concept" example in https://github.com/snowdd1/homebridge-cmd  
using a UNIX socket approach, so you can simply send a value to the switch to get its state changed in Homekit. No polling, just event driven.  
The "getValue()" of the characteristic just returns the last know state, and does not actually ask for a new state, as the characteristic value is supposed to be current, always.

Magic is in those lines https://github.com/snowdd1/homebridge-cmd/blob/master/index.js#L84 and following. You can easily apply the pattern to  
- Using a TCP socket (ip/port) instead a local UNIX socket (need to for non UNIX OSes) 
- More parsing functionality (instead of using just 0s and 1s and one socket per characteristic)
- Other characteristics. I just took the cmdswitch as basic example, because, it's just basic and simple to understand

Have a look at the last README line to see how to write values to a local UNIX socket.

Good luck!

**Comment by NorthernMan54** [comment]
@snowdd1 that is a very good example, very tight code.

@brownad I don't think your DoorBirdAccessory.prototype.getState function is working correctly, as the variable this.binaryState is never changed during events.  But after reviewing @snowdd1's example I don't think getState is needed.  Try commenting out this block 

  this.service
      .getCharacteristic(Characteristic.ProgrammableSwitchEvent)
      .on('get', this.getState.bind(this));

In getServices

Also you may want to review the DEBUG log when events are triggering, to see what messages are passed between homekit and homebridge.
